File: authandshare/src/main/java/com/fox/authandshare/auth/AuthManager.java
package com.fox.authandshare.auth;

import android.app.Activity;
import android.content.Intent;

import com.fox.authandshare.platform.PlatForm;

import java.util.HashMap;
import java.util.Map;

/**
 * @author fox.hu
 * @date 2018/8/16
 */

public final class AuthManager {
    private static final String TAG = AuthManager.class.getSimpleName();
    private Map<PlatForm, IAuth> mAuthLoginMap = new HashMap<>();
    private Map<PlatForm, AuthListener> mAuthListenerMap = new HashMap<>();
    private PlatForm mCurrentPlatForm;

    private AuthManager() {
    }

    public static AuthManager getDefault() {
        return Holder.INSTANCE;
    }

    public void fetchPlatFormInfo(PlatForm platForm, Activity activity,
            AuthListener listener) {
        IAuth iAuth = mAuthLoginMap.get(platForm);
        if (iAuth == null) {
            iAuth = AuthFactory.generate(activity, platForm);
            mAuthLoginMap.put(platForm, iAuth);
        }
        mCurrentPlatForm = platForm;
        mAuthListenerMap.put(platForm, listener);
        listener.onStart(mCurrentPlatForm);
        iAuth.fetchPlatFormInfo(activity, listener);
    }

    public void onActivityResultData(int requestCode, int resultCode, Intent data) {
        if (mCurrentPlatForm != null) {
            IAuth iAuth = mAuthLoginMap.get(mCurrentPlatForm);
            if (iAuth != null) {
                iAuth.onActivityResultData(requestCode, resultCode, data);
            }
        }
    }

    public AuthListener getIAuthListener(PlatForm platForm) {
        return mAuthListenerMap.get(platForm);
    }

    private static class Holder {
        static final AuthManager INSTANCE = new AuthManager();
    }
}

File: authandshare/src/main/java/com/fox/authandshare/auth/AuthManager.kt
package com.fox.authandshare.auth

import android.app.Activity
import android.content.Intent
import com.fox.authandshare.platform.PlatForm
import java.util.*

/**
 * @author fox.hu
 * @date 2018/8/16
 */

class AuthManager private constructor() {
    private val mAuthLoginMap = HashMap<PlatForm, IAuth>()
    private val mAuthListenerMap = HashMap<PlatForm, AuthListener>()
    private var mCurrentPlatForm: PlatForm? = null

    fun fetchPlatFormInfo(
        platForm: PlatForm, activity: Activity,
        listener: AuthListener
    ) {
        var iAuth = mAuthLoginMap[platForm]
        if (iAuth == null) {
            iAuth = AuthFactory.generate(activity, platForm)
            mAuthLoginMap[platForm] = iAuth
        }
        mCurrentPlatForm = platForm
        mAuthListenerMap[platForm] = listener
        listener.onStart(mCurrentPlatForm!!)
        iAuth!!.fetchPlatFormInfo(activity, listener)
    }

    fun onActivityResultData(requestCode: Int, resultCode: Int, data: Intent) {
        if (mCurrentPlatForm != null) {
            val iAuth = mAuthLoginMap[mCurrentPlatForm!!]
            iAuth?.onActivityResultData(requestCode, resultCode, data)
        }
    }

    fun getIAuthListener(platForm: PlatForm): AuthListener? {
        return mAuthListenerMap[platForm]
    }

    private object Holder {
        internal val INSTANCE = AuthManager()
    }

    companion object {
        private val TAG = AuthManager::class.java.simpleName

        val default: AuthManager
            get() = Holder.INSTANCE
    }
}
